main.go

```diff
@@ -48,54 +57,135 @@ func advertise() {
 
 		llog.Info("connecting to skyapi", llog.KV{"resolvedAddr": skyapiAddr, "thisAddr": config.ListenAddr})
 
-		go func() {
-			err := client.ProvideOpts(client.Opts{
-				SkyAPIAddr:        skyapiAddr,
-				Service:           "saturn",
-				ThisAddr:          config.ListenAddr,
-				ReconnectAttempts: 3,
-			})
-			llog.Fatal("skyapi giving up reconnecting", llog.KV{
-				"addr":         config.SkyAPIAddr,
-				"resolvedAddr": skyapiAddr,
-				"err":          err,
-			})
-		}()
+		err = client.ProvideOpts(client.Opts{
+			SkyAPIAddr:        skyapiAddr,
+			Service:           "saturn",
+			ThisAddr:          config.ListenAddr,
+			ReconnectAttempts: 3,
+		})
+		llog.Fatal("skyapi giving up reconnecting", llog.KV{
+			"addr":         config.SkyAPIAddr,
+			"resolvedAddr": skyapiAddr,
+			"err":          err,
+		})
+	}
+}
+
+func marshalAndWrite(msg proto.Message, c *net.UDPConn, dst *net.UDPAddr, kv llog.KV) bool {
+	b, err := proto.Marshal(msg)
+	if err != nil {
+		kv["err"] = err
+		llog.Error("error marshaling msg", kv)
+		return false
+	}
+
+	if dst == nil {
+		_, err = c.Write(b)
+	} else {
+		_, err = c.WriteToUDP(b, dst)
+	}
+
+	if err != nil {
+		kv["err"] = err
+		llog.Error("error writing msg", kv)
+		return false
 	}
+	return true
+}
+
+func readAndUnmarshal(c *net.UDPConn, kv llog.KV) (*lproto.TxMsg, *net.UDPAddr, bool) {
+	b := make([]byte, 1024)
+	n, addr, err := c.ReadFromUDP(b)
+	if err != nil {
+		kv["err"] = err
+		llog.Error("error reading from udp socket", kv)
+		return nil, nil, false
+	}
+
+	var msg lproto.TxMsg
+	if err := proto.Unmarshal(b[:n], &msg); err != nil {
+		kv["err"] = err
+		llog.Error("error unmarshaling proto msg", kv)
+		return nil, nil, false
+	}
+	return &msg, addr, true
 }
 
 func reportSpin() {
-	var masterAddr string
-	var serverAddr *net.UDPAddr
-	var err error
-	for range time.Tick(10 * time.Second) {
-		masterAddr = lookupMaster()
-		serverAddr, err = net.ResolveUDPAddr("udp", masterAddr)
-		if err != nil {
-			llog.Fatal("error resolving UDP addr", llog.KV{"err": err, "addr": masterAddr})
+	doSlaveReport()
+	for _ = range time.Tick(time.Duration(config.Interval) * time.Second) {
+		doSlaveReport()
+	}
+}
+
+func doSlaveReport() {
+	masterAddr := lookupMaster()
+	kv := llog.KV{"addr": masterAddr}
+	llog.Info("beginning transaction", kv)
+
+	c, err := net.DialUDP("udp", nil, masterAddr)
+	if err != nil {
+		kv["err"] = err
+		llog.Error("error connecting to master", kv)
+		return
+	}
+	defer c.Close()
+
+	firstMsg := transaction.Initiate()
+	kv["txID"] = firstMsg.Id
```

Comment:
The master should instead be storing the transactions by IP,id tuple instead of just id then.